### PR TITLE
Adding support for OLMoE SFT + DPO w/ and w/o load balancing loss

### PR DESCRIPTION
**Pull request by jacob-morrison**

### CODE REVIEW

File: eval/truthfulqa/run_eval.py [hunk 1]
```diff
@@ -333,6 +333,9 @@ def main(args):
             if "mc" in args.metrics:
                 raise ValueError("OpenAI Chat engines does not support MC metrics.")
 
+    del model
+    torch.cuda.empty_cache()
```
Comment (vwxyzjn) [inline eval/truthfulqa/run_eval.py:337]: oom? 👀

File: open_instruct/dpo_tune.py [hunk 2]
```diff
@@ -990,16 +1012,31 @@ def load_model():
...
+                            / args.logging_steps
+                        )
+                        logger.info(
+                            f"  Step: {completed_steps}, LR: {lr_scheduler.get_last_lr()[0]}, Loss: {avg_loss}, Aux Loss: {avg_aux_loss}"
+                        )
+                        metrics_to_log["aux_loss"] = avg_aux_loss
+                    else:
+                        logger.info(
+                            f"  Step: {completed_steps}, LR: {lr_scheduler.get_last_lr()[0]}, Loss: {avg_loss}"
+                        )
```
Comment (vwxyzjn) [inline open_instruct/dpo_tune.py]:
```suggestion
                    logger_str = f"  Step: {completed_steps}, LR: {lr_scheduler.get_last_lr()[0]}, Loss: {avg_loss}"
                    if args.load_balancing_loss:
                        avg_aux_loss = (
                            accelerator.gather(total_aux_loss).mean().item()
                            / args.gradient_accumulation_steps
                            / args.logging_steps
                        )
                        logger_str += f" Aux Loss: {avg_aux_loss}"
                        metrics_to_log["aux_loss"] = avg_aux_loss
                    logger.info(logger_str)
```

File: open_instruct/dpo_tune.py [hunk 3]
```diff
@@ -925,18 +938,23 @@ def load_model():
...
-                    model, batch, average_log_prob=average_log_prob
-                )
+                if args.load_balancing_loss:
+                    policy_chosen_logps, policy_rejected_logps, aux_loss = concatenated_forward(
+                        model, batch, average_log_prob=average_log_prob, output_router_logits=True
+                    )
+                else:
+                    policy_chosen_logps, policy_rejected_logps, _ = concatenated_forward(
+                        model, batch, average_log_prob=average_log_prob
+                    )
```
Comment (vwxyzjn) [inline open_instruct/dpo_tune.py]:
```suggestion
                policy_chosen_logps, policy_rejected_logps, aux_loss = concatenated_forward(
                    model, batch, average_log_prob=average_log_prob, output_router_logits=True
                ) # `aux_loss` is only used when `args.load_balancing_loss = True`
```

**Review by vwxyzjn — COMMENTED** [review]
Some minor nits :) Looks good!

File: eval/truthfulqa/run_eval.py (hunk 1 above)
Comment (jacob-morrison) [reply eval/truthfulqa/run_eval.py:337]: yeah, I had some ooms while evaluating deepseek's 16b MoE and I realized we can jus delete the model after this point for free wins

**Review by natolambert — APPROVED** [review]
seems fine to me!
